### ISSUE TITLE
No clear documentation on how to use registerTransformStream

### ISSUE THREAD

**Issue by mcshaman** [yeoman/yeoman#1676]
I have been searching the internet for hours now trying to find documentation on how to use the registerTransformStream() method to modify files.

The documentation is very light on and says that you can pretty much use any gulp plugins. This is great however there are no example how to pipe them in a generator.

It is recommended to use gulp-if or gulp-filter to subset any transforms... However I cannot work out how you would effectively pipe to other transformers. Do I register all the transformers in the order that I want them piped?

**Comment by SBoudrias** [comment]
@mcshaman to answer the question, yes you just call `registerTransformStream()` for each transformer in the order you want to pipe.

Let's leave this ticket open though, the documentation could be greatly improved around that point. (especially as gulp isn't used too much anymore)

**Comment by simbo1905** [comment]
I was looking for a none trivial example and agreed with this ticket. 

Eventually, I worked out the following to be able to use gulp-filter to only rewrite .json files and ignore other types. Hopefully, this will be of use to anyone else who comes across this ticket: 

```JavaScript
const jsonTransform = require('gulp-json-transform');
const filter = require('gulp-filter');
const beautify = require('gulp-beautify');

const jsonFilter = filter(['**/*.json'], { restore: true });

    //in the constructor of my Generator

    // "--scmsecret" command line flag to toggles write feature of json template
    this.option('scmsecret');

    const self = this;

    // this pipe will pass through (restore) anything that doesn't match jsonFilter
    this.registerTransformStream(
      [
        jsonFilter,
        jsonTransform(function(data, file) {
          // based on the flags passed to the generator we may or may not edit the json structure
          if( !self._private_scmsecret() ){
           // edite the json removing the scmsecret object attributes
            ...
            return data
          }
          else return data
        }),
        beautify({indent_size: 2 }),
        jsonFilter.restore
      ]
    );

```

**Comment by robblovell** [comment]
I have also been trying to use this feature to do a string replace using gulp-replace without success. It appears that when the files are being written out using this.fs.copyTpl, the writer becomes disconnected from the "this" pointer and the result of the "registerTransformStream" doesn't stick when it gets to "_writeFiles(done)" code.

Example code:
```
const Generator = require('yeoman-generator');
const chalk = require('chalk');
const replace = require('gulp-string-replace');

module.exports = class extends Generator {

  constructor(args, opts) {
    // Calling the super constructor is important so our generator is correctly set up
    super(args, opts);
    this.answers = {};
  }

  prompting() {
    return this.prompt(
      [
        {
          type: 'input',
          name: 'appname',
          message: 'What is the application name?',
          store: true,
          default: ''
        },
      ]
    ).then((answers) => {
      this.answers = {...this.answers, ...answers};
      // this.composeWith(require.resolve(`../${DEPLOYMENT_SCRIPTING}`), {this.answers});
    });
  }
  
  writing() {
    this.destinationRoot("./");

    this.registerTransformStream([
      replace("referencetriggeredjob", "yo-test")
    ]);

    this.fs.copyTpl(
      this.templatePath('reference-build-artifactory-material/*'),
      this.destinationPath(),
      this.answers
    );
  }
};
```
templates/reference-build-artifactory-material contains a file with the string "referencetriggeredjob" in it.  That file gets written out unchanged. The expected behavior is that the file will contain "yo-test".

**Comment by robblovell** [comment]
There appears to be some weirdness with the "this" pointer here.  Splicing in a hack into "_writeFiles" shows that if the transformStreams stay connected to this, the code works as expected and "thing" is replaced by "yo-test":

```
    // hack code:
    const replace = require('gulp-string-replace');

    this.registerTransformStream([
      replace("thing", "yo-test",{logs: {enabled: true}})
    ]);
    // end of hack test.
    const transformStreams = this._transformStreams.concat([conflictChecker]);
    this.fs.commit(transformStreams, () => {
      done();
    });
```

Somewhere in "run" things are getting messed up.  The run code looks pretty complex so it will take a while to learn and debug. This looks like where things might possibly go pear shaped:
```
      const writeFiles = () => {
        this.env.runLoop.add('conflicts', this._writeFiles.bind(this), {
          once: 'write memory fs to disk'
        });
      };

      this.env.sharedFs.on('change', writeFiles);
      writeFiles();
```

**Comment by robblovell** [comment]
It appears that if you use "composeWith" that you loose some connections downstream.  In the project where there is a problem, the example code that isn't working is chained like so:
```
const Generator = require('yeoman-generator');

module.exports = class extends Generator {

  constructor(args, opts) {
    // Calling the super constructor is important so our generator is correctly set up
    super(args, opts);
    this.argument('name', {type: String, required: false});
  }

  initializing() {
    if (!this.options.test) {
      this.composeWith(require.resolve('../entry-point'));
    }
  }
};
```

If the example code (in entry-point/index.js) is not linked via "composeWith", then it works as expected.

**Comment by SBoudrias** [comment]
@robblovell yeah, composed generators registered transformed aren't going to run. That was reported somewhere on https://github.com/yeoman/generator/issues (sorry I don't have time to dig it up right now). There was some context around this issue and we might have come to an agreement about what we should do about this...

**Comment by SBoudrias** [comment]
@robblovell yeah, composed generators registered transformed aren't going to run. That was reported here https://github.com/yeoman/generator/issues/819. There was some context around this issue and we might have come to an agreement about what we should do about this...

**Comment by robblovell** [comment]
Cool. I did back off from using compseWith in favor of a branching questioning strategy with my own branching organization. I was hoping to use the compse with to manage different templates as the question tree branches. I am off and running with this and I am moving forward.  I guess this issue is being moved forward on that other tree.

**Comment by exiadbq** [comment]
@robblovell I am looking for something similar with yeoman, so I setup a "demo" project(easy to run and for team contribution), then string replace all vars..Is there any update on that or it's your final solution?
Basiclly like     https://github.com/dotnet/dotnet-template-samples/blob/4af2d4d475bcd9657c0c0cee1bc352bee8b7a595/02-add-parameters/MyProject.Con/.template.config/template.json#L19

**Comment by e-dong** [comment]
FYI Just opened a new issue https://github.com/yeoman/yeoman/issues/1758, this function appears to be renamed to `queueTransformStreams`
